grpcserver/inputhandler.go

```diff
@@ -0,0 +1,119 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package grpcserver
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
+)
+
+// InputHandler wraps the input side of the shipper, and is responsible for starting and stopping the gRPC endpoint
+type InputHandler struct {
+	Shipper    ShipperServer
+	publisher  Publisher
+	log        *logp.Logger
+	server     *grpc.Server
+	startMutex sync.Mutex
+}
+
+// NewGRPCServer returns a new gRPC handler with a reference to the output publisher
+func NewGRPCServer(publisher Publisher) *InputHandler {
+	log := logp.NewLogger("input-handler")
+	srv := &InputHandler{log: log, publisher: publisher}
+
+	return srv
+}
+
+// Start runs the shipper server according to the set configuration.
+func (srv *InputHandler) Start(grpcTLS credentials.TransportCredentials, endpoint string) error {
+	//TODO: only needed while we deal with the fact that we have our config coming from multiple input units
+	if srv.server != nil {
+		srv.log.Debugf("shipper gRPC already started, continuing...")
+		return nil
+	}
+	listenAddr := strings.TrimPrefix(endpoint, "unix://")
+	var err error
+
+	srv.log.Debugf("initializing the gRPC server...")
+	opts := []grpc.ServerOption{grpc.Creds(grpcTLS)}
+	srv.server = grpc.NewServer(opts...)
+	srv.Shipper, err = NewShipperServer(srv.publisher)
+	if err != nil {
+		return fmt.Errorf("could not initialize gRPC: %w", err)
+	}
+
+	pb.RegisterProducerServer(srv.server, srv.Shipper)
+
+	srv.startMutex.Lock()
```

Comment:
Why do we need a start mutex? Why is nothing above this mutex protected?

Reply:
This is a holdover from the previous version of the server code that (I think) @rdner wrote, although this earlier comment probably explains it:
```
	// Testing that the server is running and only then unlock the mutex.
	// Otherwise if `Close` is called at the same time as `Start` it causes race condition.
```

Reply:
yes, I recall that otherwise the tests run with the elastic agent mock are flaky. There is no guarantee that `Start` is called before `Close`.